### ISSUE TITLE
InheritedPermissionFlusher should probably lazy load its services

### ISSUE THREAD

**Issue by lekoala** [silverstripe/silverstripe-framework#10851]
## Affected Version

4 & 5

## Description

The InheritedPermissionFlusher is an extension that is applied to the member (and group, but that's ok because it's cached afterwards). Currently, it's services are injected upon initialisation, but are only really needed when a flush() call is made

Since it's applied to a member, the first time you create a member instance, you get a bunch of stuff from the injector

- creating the service for the dataobjectschema
- creating a bunch of db fields instances
- and, initializing instances (global ones, and specific ones)

amongst them, the InheritedPermissionFlusher  that does a whole bunch of things:

- Creating two cache interfaces (one for the inheritedpermissions, one for the sitetreehints), which depending on your i/o can have a variable speed
- Initializing a few classes, including CMSMain, even when you are not using the CMS
- The sitetree is also initialized, which is most of the time unavoidable, but not necessary in an api context, for example

<img width="591" alt="image" src="https://github.com/silverstripe/silverstripe-framework/assets/250762/19d95b4a-7feb-43f0-a3e7-5dda6402b17f">

On my dev machine, this takes about 7ms. It's not much, but it feels wrong to have something unnecessary being loaded up on each request

on a typical setup, these services come from:

sitetreepermissions:

  SilverStripe\Security\InheritedPermissionFlusher:
    properties:
      Services:
        - '%$SilverStripe\Security\PermissionChecker.sitetree'
        - '%$SilverStripe\CMS\Controllers\CMSMain'
        - '%$SilverStripe\CMS\Model\SiteTree'

assetspermissions:

  SilverStripe\Security\InheritedPermissionFlusher:
    properties:
      Services:
        - '%$SilverStripe\Security\PermissionChecker.file'

This issue is also a reminder that any extension applied to the Member object will have direct consequences on things like SessionAuth and will have actual costs on each authenticated request.

## Solution

Instead of injecting instantiated service, it should only inject their name, and instantiate them from the injector when actually using them.

Here is with a quick POC demo, the extension instantiate for free  (=>actual gains are mostly lower than the total time shown above because some services get instantiated later anyway depending on the request)

<img width="620" alt="image" src="https://github.com/silverstripe/silverstripe-framework/assets/250762/f3388827-5436-4c2a-8de0-2e2a13018029">

and still getting the list on /dev/build as expected

<img width="269" alt="image" src="https://github.com/silverstripe/silverstripe-framework/assets/250762/c4c33f9b-c3fb-4736-9d61-8bbb8efca8f8">

**Comment by GuySartorelli** [comment]
Hi @lekoala, thank you for reporting this - and in general, thank you for all the work you've been doing lately. I may come off a bit strict sometimes (and I'm going to again in a second probably) but we really do appreciate it.

I'll be direct about this one - for a 7ms improvement we're not going to look at implementing this enhancement. And given that injection configuration is effectively public API I don't think there's any way this change could be made without waiting for a major release so a PR at this point would have nowhere to be merged.

If you can think of a way to implement a change for this that isn't a breaking change and feel strongly about this then please comment - I'll get notified so I'll be able to address your comments. But I'm going to close it for now.

**Comment by lekoala** [comment]
@GuySartorelli yes, maybe a bit of background about how I got into this would help

I was profiling the framework in general. And I was really surprised of the performance hit when getting the current member (its basically the slowest part in all the middleware init). But actually, that is expected : in the auth middleware, it's going to make a db query. And therefore, you get the initial injection of the db schema, the db fields, the data list, and other things as well. This cannot be avoided (but i do think in general that the config api is doing too much work). Most of these things will be called anyway at some point.

The most expensive part of querying the member and its first initialization is this part: the InheritPermissionFlusher. That is only used when running a flush request, which is... never, under usual circonstances.
So to summarize: most expensive part of initializing a member, is the part that is almost never used.

:-) Well, I couldn't ignore it, and the solution is easy and doesn't require an api breaking change. You can allow both lazy services and regular service injection. All it needs is adding one public getter and setter to inject services names instead of services instances. After that, it all works as expected. I think it's worth it, but I can understand that may not be a priority. I'm happy to send a PR with the changes I've made to make it work.

**Comment by GuySartorelli** [comment]
> the solution is easy and doesn't require an api breaking change

From reading the proposed solution in the issue description, it sounds like you want to change from injecting actual services to just injecting names of services - which means we have to change the configuration used from something like this (taken from `silverstripe/assets` permissions.yml):
```yml
SilverStripe\Core\Injector\Injector:
  SilverStripe\Security\InheritedPermissionFlusher:
    properties:
      Services:
        - '%$SilverStripe\Security\PermissionChecker.file'
```

to something like this:

```yml
SilverStripe\Core\Injector\Injector:
  SilverStripe\Security\InheritedPermissionFlusher:
    properties:
      ServiceNames:
        - 'SilverStripe\Security\PermissionChecker.file'
```

This is a breaking change. Any existing customisations that change _which_ services are included will simply stop working with this proposed change.

If you can show me that you have an alternative solution which doesn't include breaking changes, I'll probably be okay with it - but something like the above is a non-starter.

**Comment by michalkleiner** [comment]
My understanding is that Thomas wants to add new API to support service names instead of service instances, which would be fine in 5.1. The existing system would still work as is.

**Comment by GuySartorelli** [comment]
So long as the pre-existing service injection yaml configuration is still in place that should be fine - but aren't the pre-existing service injections the problem (rather than the lack of an alternative API for lazy-loaded injections)?

**Comment by michalkleiner** [comment]
Yeah, that might still be the issue, unless we implement it in a way that would prefer the existing injection by name, somehow. I'll leave that to @lekoala to answer :-D

**Comment by GuySartorelli** [comment]
I suppose one thing we could do is _actually_ implement a mechanism for lazyloading dependencies...

We could have a new trait that has a `getDependency()` method, and if the injector sees that method, make it _not_ auto-inject the services. And then the `getDependency()` method can be used to fetch the dependency from injector if it hasn't already been fetched, based on the properties config for that class. And then we'd need to change any existing getters to use that new method. But that still doesn't help if the existing properties aren't private because subclasses might be getting the property directly rather than using a getter method. And that's also a lot of effort for 7ms 😅

**Comment by lekoala** [comment]
@GuySartorelli  you are right, my solution indeed means that the public yml need to change. If you cannot change existing injection, it's always going to be an issue.

I still find it odd to initialize services like CMSMain on a request that doesn't use it and the current design leads to that kind of situation. I hope there will be ways to avoid these situations in v6 ;-)

**Comment by michalkleiner** [comment]
We have a few issues already prefixed "CMS 6" to make it easy to find them, why not do that here as well. We shouldn't lose value only because we don't have a `6` branch to target yet.

**Comment by GuySartorelli** [comment]
In this case I frankly don't think it's valuable _enough_ to hold the card open until the CMS 6 branch is created - since the CMS Squad certainly won't look at implementing this change, and we can't expect @lekoala or any other community member to pick it up when the time comes.

As an aside, but it's relevant since we're talking about potentially holding a card open for an extended period, the CMS Squad is going to be looking quite seriously at how we deal with stale issues and stale PRs, and there is likely to be some level of "if the issue hasn't been interacted with within X period of time, it will be closed" (we'll consult core committers closer to the time - for now we have no concrete timeline or criteria)

All of that said: If you do think this holds enough value to reopen, I won't stop you from reopening it.
